ServerCore/Helpers/NotificationHelper.cs

```diff
@@ -33,99 +23,94 @@ public NotificationHelper(ServerMessageListener messageListener)
             });
-            listeners[listener] = 0;
-        }
-
-        // Blazor is turning nulls into -1's for some reason
-        if (eventID == -1) { eventID = null; }
-        if (teamID == -1) { teamID = null; }
-        if (playerID == -1) { playerID = null; }
 
-        NotificationListener listener = new NotificationListener() { Helper = this, EventID = eventID, TeamID = teamID, PlayerID = playerID, NotificationRecieved = notificationRecieved };
-        Register(listener, listener.EventID, this.listenersByEvent);
-        Register(listener, listener.TeamID, this.listenersByTeam);
-        Register(listener, listener.PlayerID, this.listenersByPlayer);
+            listener.OnNotify += onNotify;
+        }
 
-        return listener;
-    }
-    
-    /// <summary>
-    /// Unregister from notifications. Called by the NotificationListener's Dispose().
-    /// </summary>
-    /// <param name="listener"></param>
-    private void UnregisterForNotifications(NotificationListener listener)
-    {
-        if (listener.EventID.HasValue) { this.listenersByEvent[listener.EventID.Value].Remove(listener, out _); }
-        if (listener.TeamID.HasValue) { this.listenersByTeam[listener.TeamID.Value].Remove(listener, out _); }
-        if (listener.PlayerID.HasValue) { this.listenersByPlayer[listener.PlayerID.Value].Remove(listener, out _); }
+        Register(eventID, this.listenersByEvent, onNotify);
+        Register(teamID, this.listenersByTeam, onNotify);
+        Register(userID, this.listenersByUser, onNotify);
     }
 
     /// <summary>
-    /// private NotificationListener object that holds the client's desired event handler and helps unregister when the caller disposes it.
+    /// Unregister a client from notifications
     /// </summary>
-    private class NotificationListener : IDisposable
+    /// <param name="eventID">The event the client is in</param>
+    /// <param name="teamID">The team the client is in</param>
+    /// <param name="userID">The user</param>
+    /// <param name="notificationRecieved">The EventHandler that will listen to events for the client</param>
+    public void UnregisterFromNotifications(int? eventID, int? teamID, int? userID, Func<Notification, Task> onNotify)
     {
-        public NotificationHelper Helper { get; set; }
-        public int? EventID { get; set; }
-        public int? TeamID { get; set; }
-        public int? PlayerID { get; set; }
-        public EventHandler<NotificationEventArgs> NotificationRecieved { get; set; }
-
-        public void Dispose()
+        // helper function for each of the three object types
+        void Unregister(int? id, ConcurrentDictionary<int, NotificationListener> listenerDictionary, Func<Notification, Task> onNotify)
         {
-            Helper.UnregisterForNotifications(this);
+            if (!id.HasValue) return;
+
+            NotificationListener listener = listenerDictionary.GetOrAdd(id.Value, (id) =>
```

Comment:
This could use TryGetValue to avoid creating an unnecessary object

Reply:
perils of copy/paste - thanks!